### PR TITLE
perf: use a custom probe list sync mechanism (closes #419)

### LINKED ISSUE #419: Reduce the amount of data sent through redis during fetchSockets
On the image there is data generated by a single connected probe during fetchSockets request. Most of it either rarely changes (probe info), or not required by the app logic (socket metadata).

Instead it will be nice to transfer only identificators of probes (ip and socketId for example) and all other data may be cached in memory and updated through a dedicated redis collection.

That may require to write an analogue of `@socket.io/redis-adapter` ourselves, but since fetchSockets is used almost in all app operations that may significantly improve perf.

image

### PR DESCRIPTION
**Pull request by MartinKolarik**
### Changes in the app

1. Probe list sync

   Instead of using the adapter's `fetchSockets()` based on the request/response model, we switch to a push-based system, 
where each server periodically announces its probes to other servers. Each server then composes the full list based on the 
messages it receives from other servers.

   The messages are sent once a second, and we consider the data valid for five seconds. If a server doesn't send an update in five seconds, other servers will consider it offline and remove the probes from their lists.

   The key differences of this approach are:
   1. No waiting when reading the data. All reads are immediate and return the most recent list (up to 1 second old under normal conditions, up to 5 seconds old if there are communication issues).
   2. Temporary network errors don't cause hard failures. Instead, the server works with whatever information it has available. E.g., if there are 8 servers but the communication breaks with 4 of them, the other servers will keep running measurements with roughly half the probes available instead of failing immediately. If the communication breaks completely, each server uses its local probes.
   3. Performance-wise, there are fewer messages, and they are also smaller, as we don't transfer the socket metadata that we don't need.

   As a result of this change, many places that previously worked with sockets only work with the probe data now.

   There is also a separate method that guarantees fresh data, which is only needed in the IP check.

### Changes in tests

The changes affected many tests and I've also run into some unfortunate side effects of how our tests were written and how our dependencies interacted with each other. These changes in the tests should make everything a little more robust:

1. Global fake timers

   Mocking the timers per test (suite) might work well for unit tests where everything else is mocked as well. It breaks things very badly when some of the app's code outside of the test relies on the timers because once the fake clock gets destroyed, all of its existing timers just disappear, which leads to very hard-to-debug situations.

   For scenarios where we need to control the time flow but also have the app using timers in the background, it's better to just have a single global clock that works almost like normal, except it can be forwarded as needed. This way, there are no resets or big jumps, and things don't break.

    1.1. No absolute times in tests

    As a result, we can't easily set "absolute" time in tests, as that would again have the potential to break unrelated existing timers. In most cases, these scenarios were replaced by using a relative time with the help of `relativeDayUtc` where `relativeDayUtc(10)` means "10 days from now" and `relativeDayUtc(-10)` means "10 days ago".

   1.2. Additional pause/unpause methods

   Since we can't re/create the fake clock during the tests, we also can't easily switch between `shouldAdvanceTime: true|false` setups. However, it was easy to extend the clock and add `pause()` and `unpause()` methods, which do exactly this without resetting any other state. They can be used for individual tests where precise control of the time is needed.

2. Using sandboxes instead of the global sinon instance.

   There is some cross-interaction with `testdouble`, which seems related to the fact that `testdouble` needs to reload previously loaded modules when you call `replaceEsm()`. When using the global `sinon`, it can cause the `.reset()` and `.resetHistory()` to stop working as some of the global state gets mixed up. To prevent this, I've banned the use of `sinon` methods directly in eslint, and replaced everything with "local" per-test sandboxes.

3. Extra helper methods

   There's a new `addFakeProbes(n)` function to add multiple probes at once and `waitForProbesUpdate()` to ensure any changes will be reflected in the next `getProbes()` call.

### CODE REVIEW

File: test/tests/unit/ws/reconnect-probes.test.ts [hunk 1]
```diff
@@ -30,14 +27,15 @@ describe('reconnectProbes', () => {
 	});
 
 	it('should disconnect every probe in configured time', async () => {
+		clock.pause();
```
Comment (alexey-yarmosh) [inline test/tests/unit/ws/reconnect-probes.test.ts:30]: It seems pause/unpause is not required in that test.

**Review by alexey-yarmosh — APPROVED** [review]
Nice 👍

File: test/utils/server.ts [hunk 2]
```diff
@@ -17,43 +17,79 @@ export const getTestServer = async (): Promise<Server> => {
...
+		const checker = () => {
+			if (syncedProbeList.getProbes().length >= probeCount + count) {
+				resolve(sockets);
+				syncedProbeList.off(syncedProbeList.localUpdateEvent, checker);
+			}
+		};
+
+		syncedProbeList.on(syncedProbeList.localUpdateEvent, checker);
+
+		const sockets = Promise.all(Array.from({ length: count }).map(() => new Promise<Socket>((resolve) => {
```
[the remark this answers is not in the document]
Comment (MartinKolarik) [reply test/utils/server.ts:45]: Not needed here, we're just passing the result to `resolve()` of `new Promise()`

[the remark this answers is not in the document]
Comment (alexey-yarmosh) [reply test/utils/server.ts:45]: yes, that is why it works but variable says `sockets`

[the remark this answers is not in the document]
Comment (MartinKolarik) [reply test/utils/server.ts]: True, renamed.

**Pull request merged**